File: python/WavesDisplay/comms/rabbit_subscriber.py
from comms.rabbit_client import RabbitClient


class RabbitSubscriber(RabbitClient):

    def start_consuming(self, callback):
        if self.channel is None:
            raise Exception("Not connected to RabbitMQ. Call connect() first.")

        self.channel.basic_consume(queue=self.queue, on_message_callback=callback, auto_ack=True)
        self.channel.start_consuming()

File: cpp/src/Comms/RabbitMQ/RabbitSubscriber.cpp
//
// Created by Peter on 6/24/2024.
//

#include "RabbitSubscriber.h"
#include <stdexcept>

RabbitSubscriber::~RabbitSubscriber() {
    stopConsuming();
}

void RabbitSubscriber::startConsuming(const std::function<void(const std::string &)> &callback) {
    if (!channel) {
        throw std::runtime_error("Not connected to RabbitMQ. Call connect() first.");
    }
    is_consuming = true;
    consuming_thread = std::thread([this, callback]() {
        channel->BasicConsume(queue, "", true, false, false, 1);
        while (is_consuming) {
            AmqpClient::Envelope::ptr_t envelope;
            bool success = channel->BasicConsumeMessage(envelope);
            if (success) {
                callback(envelope->Message()->Body());
            }
        }
    });
}

void RabbitSubscriber::stopConsuming() {
    if (is_consuming) {
        is_consuming = false;
        if (consuming_thread.joinable()) {
            consuming_thread.join();
        }
    }
}
